Dmitri/busy loop

a-palchikov opened the pull request:
- Update [backoff](https://github.com/cenkalti/backoff) package reference.
 - Use backoff loops w/o time restriction to avoid busy looping after the `MaxElaspedTime` has elapsed.
 - Simplify the implementation and avoid issues with potential use of null-pointers on retries
 - Use `context` where appropriate

a-palchikov committed:
Work in progress: adding proper backoff support for etcd errors

a-palchikov committed:
Revisit the watcher implementation to automatically handle renew
failures and retry indefinitely until explicitly cancelled.
Add FlippingBackOff strategy to enable combining two backoff strategies
in a loop to dynamically switch between regular update intervals and
error-driven backoff steps.

a-palchikov committed:
Add test for flipping backoff implementation and fix documentation.

a-palchikov committed:
Remove time limit in backoff loop in voter process

a-palchikov committed:
Add exponential backoff variation w/o time limit

r0mant commented on leader/backoff.go:
```diff
@@ -1,52 +1,94 @@
-// CurrentTries returns the number of attempts on this backoff counter
-func (f *FreebieExponentialBackOff) CurrentTries() int {
+// NumTries returns the number of steps taken along this backoff interval
+func (f *CountingBackOff) NumTries() int {
```
I would consider making `Tries` a public struct field and ditching this method. Or at least rename it to `Tries()` in accordance with Go style guide (https://golang.org/doc/effective_go.html#Getters).

r0mant commented on leader/backoff.go:
```diff
@@ -1,52 +1,94 @@
+
+// NewFlippingBackOff returns a new instance of the FlippingBackOff
+// using regular and failing as backoff implementations
+func NewFlippingBackOff(regular, failing backoff.BackOff) *FlippingBackOff {
```
It is not clear what "regular" and "failing" mean and which one is used in which case. Could you extend the comment?

r0mant commented on leader/backoff.go:
```diff
@@ -1,52 +1,94 @@
+
+// Failing resets the failing state to failing.
+// If failing == false, the failing backoff interval is reset.
+func (r *FlippingBackOff) Failing(failing bool) {
```
The method should be called `SetFailing`.

r0mant commented on leader/backoff.go:
```diff
@@ -1,52 +1,94 @@
+// loop that can dynamically switch between backoff implementations
+// depending on state of an operation (e.g. healthy vs experiencing transient errors).
+type FlippingBackOff struct {
+	regular, failing backoff.BackOff
```
I would split these on two lines for clarity, and added comments on each (even though they're not exported). For example, it is not very clear to me what "regular" backoff is for. Is this basically for repeating function call when it returns nil error?

a-palchikov replied:
Yes, the `regular` part is when there're no errors - e.g. something happens at regular intervals, as expected.
Poor wording, agreed.

r0mant commented on leader/leader.go:
```diff
@@ -105,15 +105,20 @@ func (l *Client) AddWatchCallback(key string, retry time.Duration, fn CallbackFn
+	}
+	if resp == nil {
+		log.Debug("client is closing")
 		return nil, nil, nil
```
I know this is not a part of your changes, but `nil, nil, nil` is a pretty bad return value as it can potentially lead to a panic if the caller expects first values to be usable in case of nil error. If we changed this to return an error, do you know if this would require a lot of refactoring in code that uses this?

a-palchikov replied:
The caller does not expect the value to be non-nil (see comment below).

r0mant commented on leader/leader.go:
```diff
@@ -124,229 +129,218 @@ func (l *Client) getWatchAtLatestIndex(ctx context.Context, api client.KeysAPI,
+			log.Debug("electing")
+			err := l.elect(ctx, key, value, term)
+			if err != nil {
+				log.Debugf("voter error: %v", err)
```
`trace.DebugReport(err)`

r0mant commented on leader/leader.go:
```diff
@@ -124,229 +129,218 @@ func (l *Client) getWatchAtLatestIndex(ctx context.Context, api client.KeysAPI,
+		return
+	}
+
+	// send sends the response value on the changes channel.
```
sends _to_ the channel

r0mant commented on leader/leader.go:
```diff
@@ -124,229 +129,218 @@ func (l *Client) getWatchAtLatestIndex(ctx context.Context, api client.KeysAPI,
+			watcher, resp, err = l.recreateWatchAtLatestIndex(ctx, api, key, retry)
+			return trace.Wrap(err)
+		}, b)
+		return watcher, resp
```
I would make this function return an error in case watcher could not be created. Right now it just returns nils as far as I understand.

a-palchikov replied:
Yes, it omits the error on purpose as the retry loop is not time-limited and the `watcher == nil` is special in the sense that it indicates the event of closing the client.
The whole retry is only supposed to be cancelled (via a context or calling `client.Close`) - this is all due to the way API is structured (e.g. all errors are handled internally by the `leader` package so it makes sense to forego error handling at certain points).

The alternative is moving the error handling out of the leader loops into the client (analogue to https://github.com/docker/leadership, which has a much simpler implementation as result).

a-palchikov committed:
Address review comments

a-palchikov committed:
Remove logging redundancies and properly use retry as an initial
interval for fetching values from watcher.

klizhentas commented on leader/backoff.go:
```diff
@@ -1,52 +1,100 @@
-	MaxElapsedTime      time.Duration
+// NewCountingBackOff returns a new instance of the CountingBackOff
+// with the specified backoff as implementation
+func NewCountingBackOff(backOff backoff.BackOff) *CountingBackOff {
```
this is all too complex, can we replace all this thing with something else, even static 1 second delay + randomizer will work here.

klizhentas commented on leader/leader.go:
```diff
@@ -105,15 +105,17 @@ func (l *Client) AddWatchCallback(key string, retry time.Duration, fn CallbackFn
 	}
-	log.Debugf("%v got current value '%v' for key '%v'", ctx.Value("prefix"), resp.Node.Value, key)
+
+	entry.Debugf("got current value %q for key %q", resp.Node.Value, key)
```
this is super verbose to be honest, and our logs are filled with it, can we remove this line

klizhentas commented on leader/leader.go:
```diff
@@ -124,229 +126,222 @@ func (l *Client) getWatchAtLatestIndex(ctx context.Context, api client.KeysAPI,
 				return
 			}
+		case <-ticker.C:
+			log.Debug("electing")
```
this will pollute the logs

klizhentas commented on leader/leader.go:
```diff
@@ -124,229 +126,222 @@ func (l *Client) getWatchAtLatestIndex(ctx context.Context, api client.KeysAPI,
+			log.Debug("electing")
+			err := l.elect(ctx, key, value, term)
+			if err != nil {
+				log.Debugf("voter error: %v", trace.DebugReport(err))
```
Debugf and error?

klizhentas commented on leader/leader.go:
```diff
@@ -124,229 +126,222 @@ func (l *Client) getWatchAtLatestIndex(ctx context.Context, api client.KeysAPI,
+		case <-l.closeC:
+			return
+		case <-ctx.Done():
+			log.Debug("removing voter")
```
this line is not helpful? which voter

a-palchikov replied:
`log` is scoped to the value of the watched key (see above)

klizhentas commented on leader/leader.go:
```diff
@@ -124,229 +126,222 @@ func (l *Client) getWatchAtLatestIndex(ctx context.Context, api client.KeysAPI,
 			return trace.Wrap(err)
 		}
-		log.Debugf("%v key not found, try to elect myself", candidate)
+		log.Debug("key not found, try to elect myself")
```
try to elect who?

a-palchikov replied:
`log` is scoped to the candidate running the election

klizhentas commented on leader/leader.go:
```diff
@@ -124,229 +126,222 @@ func (l *Client) getWatchAtLatestIndex(ctx context.Context, api client.KeysAPI,
 	}
 	if resp.Node.Value != value {
-		log.Debugf("%v: leader is %v, try again", candidate, resp.Node.Value)
+		log.Debugf("leader is %q, try again", resp.Node.Value)
```
this line is polluting the logs

a-palchikov committed:
Address review comments

a-palchikov committed:
Drop another log entry

a-palchikov committed:
Properly terminate renew function if either the context has been
canceled or client closed.
Factor decision whether send the value into a separate function.

a-palchikov committed:
Revisit renewFunc to properly return error on failure to create a watch.
Handle context error consistently.

klizhentas commented on leader/backoff.go:
```diff
@@ -1,52 +1,67 @@
+	r.isFailing = failing
+	if !failing {
+		r.failing.Reset()
+	}
```
why don't you reset regular backoff here if it's `failing==true`?

klizhentas commented on leader/leader.go:
```diff
@@ -124,229 +121,239 @@ func (l *Client) getWatchAtLatestIndex(ctx context.Context, api client.KeysAPI,
-	go func() {
+	entry := log.WithFields(log.Fields{"watch": key})
+	renewCtx, cancel := context.WithCancel(ctx)
+	renew := func() (client.Watcher, *client.Response, error) {
```
this code is hard to read - gorotuine and a callback inside with inversion of control? can you simplify this? I don't understand by we need this complex backoff/callback system here.

klizhentas commented on leader/leader.go:
```diff
@@ -124,229 +121,239 @@ func (l *Client) getWatchAtLatestIndex(ctx context.Context, api client.KeysAPI,
-			Multiplier:          1.5,
-			MaxInterval:         10 * time.Second,
-			MaxElapsedTime:      30 * time.Second,
+		err := backoff.Retry(func() (err error) {
```
this line is really hard to read, I also don't understand why it is necessary to use callbacks here

klizhentas commented on leader/leader.go:
```diff
@@ -124,229 +121,239 @@ func (l *Client) getWatchAtLatestIndex(ctx context.Context, api client.KeysAPI,
 
-		watcher, resp, err = l.getWatchAtLatestIndex(ctx, api, key, retry)
-		if err != nil {
+	go func() {
```
why is this a separate goroutine?

klizhentas commented on leader/leader.go:
```diff
@@ -124,229 +121,239 @@ func (l *Client) getWatchAtLatestIndex(ctx context.Context, api client.KeysAPI,
+	go l.watcher(ctx, renew, valuesC, entry)
+}
+
+func (l *Client) watcher(ctx context.Context, renew renewFunc, valuesC chan string, entry *log.Entry) {
```
this is a very complex logic, and became even more complex. Why all these changes?

a-palchikov commented:
Closing in favor of a simpler solution.
